### PR TITLE
test: add reusable routed forwarding SLO harness

### PR DESCRIPTION
**Pull request by lucas-albers-lz4**
## Summary

- add reusable two-endpoint QEMU routed topology setup and fail-closed teardown
- add MAC-selected guest forwarding/logging setup and cleanup
- add routed iperf3/ping probe plus real LuCI viewer worker
- add paired adaptive-on/off runner with request drain, cadence, raw samples, medians, spread, and versioned JSON reports
- document the harness reuse contract and retain security-review coverage

## Validation

- `./scripts/fwlive-test.sh` passes
- armsr TCG 1G five-pair runs pass the harness validity, throughput, and RTT gates for adaptive on/off
- actual loaded polling cadence is recorded as evidence (about 3–5 seconds in that environment)

## Qualification status

The harness and benchmark infrastructure are complete for reuse. The forwarding qualification remains open pending review of the loaded cadence and the issue-defined x86_64 validation scope. The armsr TCG results are documented as qualified lab evidence, not native ARM performance claims.

Refs #306
Refs #344

## Summary by CodeRabbit

* **New Features**
  * Added a forwarding SLO lab harness for routed topology setup, guest forwarding rules, traffic probes, and paired viewer/no-viewer performance runs.
  * Added configurable NIC models, adaptive-mode handling, ping intervals, bitrate settings, and secure report output.
  * Added JSON reporting with throughput, latency, polling, request-drain, and acceptance criteria, including optional enforcement mode.
  * Added a viewer monitor for tracking polling activity and request health.

* **Documentation**
  * Documented topology setup, measurement parameters, cleanup behavior, and report validity gates.

* **Tests**
  * Added automated validation for the network, harness, viewer, and report workflows.

### CODE REVIEW

**Comment by lucas-albers-lz4** [comment]
# Multi-Model Review — lucas-albers-lz4/fwlive PR #345

> **Head SHA:** `60d1ee3f32`  •  **Reviewers:** big-pickle, nemotron-3-ultra-free, mimo-v2.5-free  •  **Synthesized by:** muse
> **Verdict:** COMMENT

## TL;DR
This PR adds a reusable two-endpoint QEMU forwarding-SLO lab harness (host netns/tap topology, guest forwarding/logging rules, routed iperf3/ping probe, Playwright viewer worker, paired adaptive on/off runner with versioned `fwlive-forwarding-slo/v1` JSON reports) plus docs and static tests. All three reviewers agree the structure, fail-closed teardown, validation, and evidence schema are high quality. The orchestrator-verified ground truth confirms **no Critical defects**, but confirms five Major-grade issues (drain-sleep arithmetic for `--drain-ms < 100`, viewer drain-gate flakiness, untested report/acceptance block, MAC-override mismatch, shellcheck gap omitting the core lib) plus a set of Minors/nits. Three reviewer claims were proven false by execution and are excluded below. Default documented usage (5 pairs, 10s, 3s drain) is unaffected, so COMMENT — fix the drain/MAC items before relying on enforcement mode.

## Findings (sorted by severity)

### 🔴 Critical
None confirmed. Reviewer B's Critical #2 (awk `$2`/`$4` ping-parse) was proven FALSE by executing the exact awk (see excluded list). No other reviewer claimed a Critical.

### 🟡 Major
- **Finding:** Drain-sleep arithmetic wrong for `--drain-ms < 100`
- **File:** `scripts/qemu-forwarding-slo-run.sh:154`
- **Reviewers who flagged:** B, A
- **Problem:** `sleep "$((DRAIN_MS / 1000)).$((DRAIN_MS % 1000))"` builds an unpadded decimal. Default 3000 → `3.0` (correct) and 500 → `0.500` (correct), but 50 → `0.50` parsed by GNU sleep as 0.5s = 500ms (10x off) and 5 → `0.5` = 500ms (100x off). Paired-comparison timing silently stretches for small drain values.
- **Suggested fix:** `printf "%d.%03d" $((DRAIN_MS/1000)) $((DRAIN_MS%1000))` or `sleep "${DRAIN_MS}ms"` (GNU). Downgraded from B-Critical to Major per digest: default path unaffected.
- **Confidence:** high
- **Finding:** Viewer drain gate is flaky — counts in-drain polls as in-flight
- **File:** `tests/fwlive-forwarding-slo-viewer.mjs:139-153,185-190` (gate at `scripts/qemu-forwarding-slo-run.sh:251`)
- **Reviewers who flagged:** A
- **Problem:** `inFlight.add(request)` fires for every fwlive request unconditionally, and nothing stops the page's polling after the `stop` marker (only `browser.close()` in `finally`). Polls issued during the 3s post-stop drain can still be unresolved when `in_flight_after_drain` is sampled, so `viewer_requests_drained === 0` can spuriously fail/pass.
- **Suggested fix:** Clear `inFlight` at the measure boundary, or stop page polling (navigate/close) before draining, and count only pre-stop requests.
- **Confidence:** medium
- **Finding:** Report/stats/acceptance Node block has no behavioral test
- **File:** `scripts/qemu-forwarding-slo-run.sh:196-274` (greps only in `tests/qemu-forwarding-slo-harness.test.sh:30-41`)
- **Reviewers who flagged:** A
- **Problem:** ~70-line medians/pair-deltas/completeness/validity/`pass` logic (edge cases: `ping_stddev_ratio: null`, incomplete pairs, empty medians) is covered only by literal substring greps; easy to regress silently.
- **Suggested fix:** Extract the block to a module and unit-test against synthetic `records.jsonl` fixtures with the runner's exact input shapes.
- **Confidence:** high
- **Finding:** MAC env overrides silently break QEMU↔guest matching
- **File:** `scripts/lib/qemu-forwarding-slo-net.sh:185,187` vs `scripts/qemu-forwarding-slo-guest.sh:19-20`
- **Reviewers who flagged:** A
- **Problem:** `fwlive_slo_net_qemu_args` hardcodes `52:54:00:30:77:01/02` while the guest helper accepts `FWLIVE_SLO_LAN_MAC`/`WAN_MAC` overrides. An override makes guest `find_dev` unmatchable with no cross-check, contradicting the documented "use environment overrides for MACs" contract.
- **Suggested fix:** Read the MAC from the same env vars in `qemu_args`, or validate overrides match the printed TAP MACs.
- **Confidence:** medium
- **Finding:** Shellcheck omits the core topology library
- **File:** `tests/qemu-forwarding-slo-harness.test.sh:21-23`
- **Reviewers who flagged:** B
- **Problem:** Guest/traffic/run scripts are linted but `scripts/lib/qemu-forwarding-slo-net.sh` (189 lines, core setup/teardown/validation) is not, leaving the highest-risk shell unlinted in CI.
- **Suggested fix:** Add `"$ROOT/scripts/lib/qemu-forwarding-slo-net.sh"` to the shellcheck command line.
- **Confidence:** high

### 🟢 Minor
- **Finding:** nft cleanup can orphan rules on mid-list failure; `check` is weak
- **File:** `scripts/qemu-forwarding-slo-guest.sh:72-80,113`
- **Reviewers who flagged:** A, C
- **Problem:** Deletes handles in a `while read` subshell under remote `set -eu`; a mid-list delete failure aborts the loop leaving orphans. `check` passes if any `fwlive-slo-*` substring matches even when the two accept rules are gone. C's renumbering mechanism is false (nft handles are stable, unlike iptables line numbers) — corrected per digest; the error-path orphaning is the real residual.
- **Suggested fix:** Collect handles first then delete in a loop with a final existence re-check; make `check` require both accept rules.
- **Confidence:** medium
- **Finding:** WAN veth brought up twice; LAN veth `up` missing in master block
- **File:** `scripts/lib/qemu-forwarding-slo-net.sh:115`
- **Reviewers who flagged:** A
- **Problem:** Copy/paste slip; masked only by the final restore at `:134-135`. No functional impact today.
- **Suggested fix:** Change line 115 to target `$FWLIVE_SLO_LAN_VETH`.
- **Confidence:** high
- **Finding:** `--help` on viewer requires full playwright install
- **File:** `tests/fwlive-forwarding-slo-viewer.mjs:9-10` (invoked at `tests/qemu-forwarding-slo-harness.test.sh:28`)
- **Reviewers who flagged:** A
- **Problem:** Top-level `import './lib/playwright-lab.mjs'` → playwright means `node --help` needs `node_modules`; fails with `ERR_MODULE_NOT_FOUND` on a fresh checkout. Passes here only because `node_modules` is present.
- **Suggested fix:** Move the playwright import inside `main()` or drop the live `--help` invocation.
- **Confidence:** high
- **Finding:** `expect_absent` skips veth/peer names
- **File:** `scripts/lib/qemu-forwarding-slo-net.sh:92-97`
- **Reviewers who flagged:** A
- **Problem:** Only netns/bridge/tap names are pre-flight checked; the four veth/peer names (in `validate_names` and teardown) are not, so a stale half-pair is salvaged only by the ERR-trap teardown.
- **Suggested fix:** Add the four veth/peer names to the `expect_absent` loop.
- **Confidence:** medium
- **Finding:** MAC validation regex duplicated for LAN/WAN
- **File:** `scripts/qemu-forwarding-slo-guest.sh:26-33`
- **Reviewers who flagged:** B
- **Problem:** Maintenance burden; inconsistency risk.
- **Suggested fix:** Extract to a `validate_mac()` function.
- **Confidence:** high
- **Finding:** iperf3 search path includes macOS-linuxbrew-specific path
- **File:** `scripts/qemu-forwarding-slo-traffic.sh:66-73`
- **Reviewers who flagged:** B
- **Problem:** `/home/linuxbrew/.linuxbrew/bin/iperf3` is platform-specific; confusing elsewhere.
- **Suggested fix:** Remove it; rely on standard locations plus `FWLIVE_SLO_IPERF3` override.
- **Confidence:** medium
- **Finding:** Name validation allows the bare prefix as a resource name
- **File:** `scripts/lib/qemu-forwarding-slo-net.sh:46-53`
- **Reviewers who flagged:** B
- **Problem:** `"${PREFIX}"[A-Za-z0-9.-]*|"${PREFIX}"` accepts `fwlive-slo-` alone.
- **Suggested fix:** Require at least one char after prefix (`+` instead of `*`, drop the bare alternative).
- **Confidence:** medium
- **Finding:** `harness.test.sh` not executable (644 vs 755)
- **File:** `tests/qemu-forwarding-slo-harness.test.sh:1`
- **Reviewers who flagged:** C
- **Problem:** Inconsistent with `net.test.sh` (755). No functional impact (`fwlive-test.sh` invokes via `bash`).
- **Suggested fix:** `chmod +x tests/qemu-forwarding-slo-harness.test.sh`.
- **Confidence:** high
- **Finding:** `ns_exists` parsing is fragile
- **File:** `scripts/lib/qemu-forwarding-slo-net.sh:75-77`
- **Reviewers who flagged:** C
- **Problem:** `ip netns list | awk '{print $1}' | grep -Fxq` can false-negative on format/space edge cases.
- **Suggested fix:** Use `ip netns list -n` name-only output where available, falling back to current behavior.
- **Confidence:** low

### 💡 Nits
- **Finding:** `request_failures` counts pre-window failures (incl. login)
- **File:** `tests/qemu-forwarding-slo-viewer.mjs:151-153`
- **Reviewers who flagged:** A
- **Problem:** `requestfailed` handler not gated on `measureStarted`; slightly conflates the reported metric.
- **Suggested fix:** Gate increment on measurement window.
- **Confidence:** high
- **Finding:** Drain sleep also fires before pair 1's baseline
- **File:** `scripts/qemu-forwarding-slo-run.sh:154`
- **Reviewers who flagged:** A
- **Problem:** Unnecessary 3s before the first baseline; harmless.
- **Suggested fix:** Skip drain before pair 1.
- **Confidence:** medium
- **Finding:** Viewer timeout uses hardcoded 60s buffer
- **File:** `scripts/qemu-forwarding-slo-run.sh:169` (`DURATION*1000+60000`)
- **Reviewers who flagged:** B
- **Problem:** Magic number; may be wrong for very slow/fast environments.
- **Suggested fix:** Make buffer configurable or derive from DRAIN_MS/PING_COUNT.
- **Confidence:** low
- **Finding:** Ping interval hardcoded `-i 0.5`
- **File:** `scripts/qemu-forwarding-slo-traffic.sh:98`
- **Reviewers who flagged:** B
- **Problem:** If DURATION/PING_COUNT change, ping window may not fit inside iperf window.
- **Suggested fix:** Calculate interval as `DURATION / PING_COUNT` with margin.
- **Confidence:** low
- **Finding:** Fixed 1s sleep after iperf server start
- **File:** `scripts/qemu-forwarding-slo-traffic.sh:90-93`
- **Reviewers who flagged:** C
- **Problem:** Wastes ~10s over 10 samples on fast systems; may be insufficient on slow ones. Lab-only.
- **Suggested fix:** No change required; future: poll server port.
- **Confidence:** medium

## Reviewer agreement matrix

| Finding | A | B | C | Notes |
|---|---|---|---|---|
| Drain-sleep arithmetic | ✓ (Minor) | ✓ (Critical) | — | Digest: Major; default path safe, `<100ms` 10–100x off |
| Viewer drain-gate flakiness | ✓ | — | — | Confirmed, medium conf |
| Report block untested | ✓ | — | — | Confirmed |
| MAC-override mismatch | ✓ | — | — | Confirmed, medium conf |
| Shellcheck omits lib | — | ✓ | — | Confirmed, Major-grade gap |
| nft cleanup orphan / weak check | ✓ (accurate form) | — | ✓ (wrong renumber mechanism) | Mechanism corrected per digest; residual is Minor |
| WAN veth double-up | ✓ | — | — | Confirmed cosmetic |
| `--help` needs playwright | ✓ | — | — | Conditional (passes when `node_modules` present) |
| `expect_absent` misses veth/peers | ✓ | — | — | Single-reviewer Minor |
| MAC regex duplicated | — | ✓ | — | Single-reviewer Minor |
| Linuxbrew iperf path | — | ✓ | — | Single-reviewer Minor |
| Bare-prefix name accepted | — | ✓ | — | Single-reviewer Minor |
| harness.test.sh not executable | — | — | ✓ | Verified `ls -la`; no functional impact |
| `ns_exists` fragile | — | — | ✓ | Low conf |
| `request_failures` pre-window | ✓ (nit) | — | — | Confirmed |
| Drain before pair 1 | ✓ (nit) | — | — | Harmless |
| Viewer 60s buffer hardcoded | — | ✓ (nit) | — | Low conf |
| Ping `-i 0.5` hardcoded | — | ✓ (nit) | — | Low conf |
| Fixed 1s server sleep | — | — | ✓ (nit) | Lab-only |
| awk `$2` ping-parse bug (B-Crit #2) | — (implicit ✗: A praised report null-safety) | ✓ | — | **Excluded: proven FALSE by digest (exec test)** |
| iperf3 stderr lost (C-Maj #2) | — | — | ✓ | **Excluded: proven FALSE (traffic.sh:107-110 cats errs)** |
| `set_adaptive` swallowed (C-Nit) | — | — | ✓ | **Excluded: proven FALSE (caller under `set -e`, aborts)** |

(✓ = flagged, — = didn't flag, ✗ = explicitly said fine)

## What the reviewers disagreed about
- **Severity of the drain-sleep bug:** B called it Critical, A called the same line cosmetic Minor, digest splits the difference at Major (correct only for `--drain-ms < 100`; default 3000 unaffected). Adopted as Major.
- **nft cleanup mechanism:** C claimed handles renumber after each delete (like iptables); digest proves this FALSE — nft handles are stable. A's error-path framing (mid-list failure under `set -eu` + weak `check`) is the correct residual. Downgraded from C-Major to Minor.
- **Whether the traffic probe is broken:** B-Critical claimed ping-stddev extraction always fails; C-Major claimed diagnostics are lost on iperf failure. Both proven FALSE by digest (awk `-F=` makes `$2` the values field; `client.err/server.err/ping.err` are catted). No change needed.
- **Whether `set_adaptive` needs `|| die`:** C suggested it; digest shows the caller is under `set -euo pipefail` so SSH failure already aborts. Redundant but harmless — excluded.
- No other substantive disagreements; all three agree the harness structure, validation, and versioned evidence schema are good.

## Things the reviewers may have missed
- **Synthesizer-only (low, Minor): `configure` uses `nft insert` ×4, so final chain order is reverse of listing.** Both log rules (which also `accept`) end up before the plain accept rules. Functionally equivalent (all accept) but counters/log split is confusing; consider `add` or explicit positions with a comment. Confidence: low — needs lab `nft -a list` to confirm ordering matters for counters.
- **Synthesizer-only (low, Nit): `BITRATE` regex `^[0-9]+([KMGkmg])?$` rejects decimal rates (`1.5G`) and `iperf3`-style suffixes (`100M` ok, `10m` ok, `1G` ok).** Document the accepted grammar or widen to `^[0-9]+(\.[0-9]+)?[KMGkmg]?$`. Confidence: low.
- Otherwise none beyond the above — reviewers covered validation, teardown, error paths, and test gaps well. No new security/performance issue found on own diff read (prefix-scoped teardown, 0600 reports, absolute iperf3 path, bounded numeric inputs all hold).

## Suggested next actions for the author
- [ ] Fix drain-sleep formatting (`printf "%d.%03d"` or `"${DRAIN_MS}ms"`) and skip the pre-pair-1 drain
- [ ] Fix viewer drain gate: snapshot/clear `inFlight` at stop boundary or stop page polling before draining; gate `request_failures` on `measureStarted`
- [ ] Fix MAC contract: derive QEMU TAP MACs from `FWLIVE_SLO_LAN_MAC`/`WAN_MAC` or cross-check overrides
- [ ] Add `scripts/lib/qemu-forwarding-slo-net.sh` to shellcheck in `harness.test.sh`; add behavioral unit test for the report/acceptance block with synthetic `records.jsonl` fixtures
- [ ] Harden guest cleanup: collect-then-delete with final re-check; require both accept rules in `check`
- [ ] Polish Minors: fix LAN/WAN veth `up` slip; lazy-load playwright import for `--help`; extend `expect_absent` to veth/peers; reject bare-prefix names; dedupe MAC regex; drop linuxbrew path; `chmod +x` harness test

## Reviewer confidence summary
- Reviewer A: High — thorough, line-precise, fail-closed/validation/schema praise checks out; 3 Majors + 4 Minors all verified TRUE (sleep item understated as Minor but same root cause as B's catch).
- Reviewer B: Mixed — 1 real Major (sleep arithmetic, overstated as Critical) + 1 real Major-grade gap (shellcheck) + 3 valid Minors; but Critical #2 (awk) proven FALSE by execution, indicating no `-F=` analysis.
- Reviewer C: Mixed — 1 valid Minor (mode 644) + 1 valid Nit + accurate instinct on cleanup fragility, but Major #1 mechanism (renumbering) FALSE, Major #2 (stderr) FALSE, and Nit #1 (`|| die`) redundant per `set -e`.

## Process notes
- Wall time: 180 seconds
- Reviewer failures: none
- Diff was truncated: no
- Total cost: $0 (Zen free tier)
- Verification: digest attached; false-positive reviewer claims excluded per ground truth

**Comment by lucas-albers-lz4** [comment]
Review follow-up for e568d27:

- Restored pre-run guest forwarding/link/address state and the adaptive sentinel state.
- Corrected viewer request completion/drain accounting and bounded measurement to the actual traffic window.
- Captured run-start time, corrected even-sample medians, extracted and unit-tested report/acceptance logic, made request failures fail closed, and published reports through a private atomic 0600 staging file.
- Hardened topology rollback/preflight/name validation, MAC/model overrides, exact nft rule checks/cleanup, and included the core topology library in ShellCheck.
- Added dynamic ping timing and the x86 QEMU management-NIC model override needed to keep management separate from the routed test NICs.
- Validation passed: full ./scripts/fwlive-test.sh, targeted shell/Node checks, and a real x86 one-pair smoke run with guest configure/check/cleanup restoration.

The remaining CodeRabbit event is a plan rate-limit notice; no new actionable comments were posted. Qualification benchmarking remains separate from this implementation/review pass.

**Comment by lucas-albers-lz4** [comment]
@coderabbitai review

**Comment by lucas-albers-lz4** [comment]
Post-review robustness fix pushed as a0350c3: the active-traffic failure path now only creates the stop marker if it is absent, because the sudo traffic helper may already have created a root-owned marker. This preserves viewer shutdown without an unwritable-marker error. Targeted harness checks and the full ./scripts/fwlive-test.sh pass.\n\nBenchmark evidence is recorded on #344: both adaptive-on and adaptive-off five-pair 100M runs passed the report gates; a separate 1G smoke correctly failed closed on this x86 QEMU path after packet loss and is not qualification evidence.

**Comment by lucas-albers-lz4** [comment]
Addressed the nine actionable findings from the latest CodeRabbit round in 5695d9d:\n\n- Fully restricted owned resource suffix validation and moved QEMU NIC-model validation to QEMU-argument emission.\n- Added EXIT-based topology rollback for partial setup and preserved guest state files when either rollback operation fails.\n- Tightened the nft logging-prefix check to require the exact installed prefix.\n- Made adaptive-sentinel restoration and temporary-work cleanup failures visible and status-preserving.\n- Added a bounded pre-navigation viewer drain and excluded teardown cancellations from request-failure counts.\n- Required one valid ping ratio per requested pair before the ratio gate can pass.\n- Staged reports privately in the destination directory for same-filesystem atomic replacement, with updated symlink-safe tests.\n\nValidation: targeted shell/Node/ShellCheck checks and the full ./scripts/fwlive-test.sh pass. The PR remains Ready for the next automatic CodeRabbit review.

File: tests/fwlive-forwarding-slo-viewer.mjs [hunk 1]
```diff
@@ -0,0 +1,254 @@
...
+			requests: Object.fromEntries(
+				Object.entries(methodCounts)
+					.filter(([method]) => method.startsWith('window_'))
+					.map(([method, count]) => [method.slice('window_'.length), count])
+			),
+			polls_in_window: requestTimes.length,
+			poll_cadence_ms: intervalsSummary(requestTimes),
+			request_failures: requestFailures,
+			in_flight_at_window_end: requestsBeforeDrain,
+			in_flight_after_drain: measuredInFlight.size,
```
Comment (lucas-albers-lz4) [inline tests/fwlive-forwarding-slo-viewer.mjs]: `in_flight_after_drain` is sampled after leftovers are ignored and `page.goto('about:blank')` cancels them. `viewer_requests_drained` / `--enforce` can therefore pass when requests did not actually settle within `--drain-ms`.

The reuse contract says a run is evidence only when the active workload has zero in-flight requests after drain. This metric can satisfy that by cancelling, not completing. `in_flight_at_window_end` is already recorded and unused.

Suggested fix:
1. Snapshot `measuredInFlight.size` immediately after the first drain-wait loop, before `ignoredFailures` / navigation.
2. Keep the ignore set only for `request_failures` so teardown cancel does not inflate failures.
3. Add a report fixture where `in_flight_after_drain !== 0` fails `pass`.

**Review by lucas-albers-lz4 — COMMENTED** [review]
## Review (worktree at 5695d9d)

Reusable lab harness looks solid after the review-fix commits: prefix-scoped teardown, MAC-selected guest links, saved forwarding/link/adaptive restore, marker-synced viewer window, and unit-tested report gates.

One remaining issue before `--enforce` is a trustworthy drain-validity signal. Snapshot `in_flight_after_drain` after the drain wait and before `about:blank` cancel. Details on the viewer worker.

Not defects, for awareness only:
- `summary_mode_seen` should stay observational (adaptive-off omits `summary`)
- `docs/developer/qemu-lab.md` still has a future-tense “will add … in #344” paragraph above the working run commands
- a baseline `ping_rtt_stddev_ms === 0` fail-closes `--enforce` (correct, but can bite a very stable TCG lab)

**Comment by lucas-albers-lz4** [comment]
Addressed the remaining drain-validity concern in e8c9027. The viewer now records in_flight_after_drain at the bounded natural-drain deadline before about:blank teardown; requests canceled by teardown remain excluded from request_failures, but they can no longer make the drain gate appear clean. Targeted checks and the full ./scripts/fwlive-test.sh pass.

**Comment by lucas-albers-lz4** [comment]
@coderabbitai review

**Comment by lucas-albers-lz4** [comment]
Astra general review follow-up: both P2 findings were addressed in 35f7d61 (successful RPC/payload validation and forwarding custom topology overrides through sudo). Full ./scripts/fwlive-test.sh passes, all PR checks are green, and CodeRabbit posted no additional findings (review was rate-limited). Proceeding with squash merge.
